### PR TITLE
Dhis2 10186 nti get enrollments

### PR DESCRIPTION
**Pull request by gnespolino**

### CODE REVIEW

File: dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/tracker/export/TrackerEnrollmentsExportController.java [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+ *  ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ *  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ *  ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ *  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ *  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+package org.hisp.dhis.webapi.controller.tracker.export;
```
Comment (ameenhere) [inline dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/tracker/export/TrackerEnrollmentsExportController.java:30]: package declaration before copyright

File: dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/service/ContextService.java [hunk 2]
```diff
@@ -76,10 +77,17 @@
...
      * Get all parameters as a map of key => values, supports more than one pr key (so values is a collection)
      */
     Map<String, List<String>> getParameterValuesMap();
+
+    /**
+     * Get a list of fields from request
+     */
+    List<String> getFieldsFromRequestOrAll();
+
+    List<String> getFieldsFromRequestOrElse(String s);
```
Comment (ameenhere) [inline dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/service/ContextService.java:92]: formatting, space paranthesis

File: dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/service/TrackedEntityInstanceSupportService.java [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+ *  ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ *  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ *  ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ *  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ *  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+package org.hisp.dhis.webapi.service;
```
Comment (ameenhere) [inline dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/service/TrackedEntityInstanceSupportService.java:30]: package declaration before copyright

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/trackedentity/AbstractTrackedEntityInstanceService.java [hunk 4]
```diff
@@ -200,61 +200,8 @@
     }
 
     @Override
-    @Transactional( readOnly = true )
```
Comment (ameenhere) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/trackedentity/AbstractTrackedEntityInstanceService.java:203]: I see that the previous getTrackedEntityInstances2() method did not have the @Transactional (readOnly = true) declaration. Would it make sense to add it?

**Review by ameenhere — APPROVED** [review]
Minor comments.

**Review by enricocolasante — COMMENTED** [review]
Can you rebase it with master?
I think there are a lot of things that were already reviewed

**Comment by gnespolino** [comment]
@enricocolasante yep, I see, the point is that I branched this from TrackedEntities GET endpoint branch since it wasn't already merged in master and I needed some common stuff to work on this. Now that branch is merged so basically PR is against master (old source branch doesn't exists anymore). I'll try to pull from master maybe probably the diff will become more effective, or I can create a new PR cherry picking only relevant commits from this one.

**Comment by enricocolasante** [comment]
> @enricocolasante yep, I see, the point is that I branched this from TrackedEntities GET endpoint branch since it wasn't already merged in master and I needed some common stuff to work on this. Now that branch is merged so basically PR is against master (old source branch doesn't exists anymore). I'll try to pull from master maybe probably the diff will become more effective, or I can create a new PR cherry picking only relevant commits from this one.

Just a merge and the common commits should disappear from this PR

**Comment by gnespolino** [comment]
> > @enricocolasante yep, I see, the point is that I branched this from TrackedEntities GET endpoint branch since it wasn't already merged in master and I needed some common stuff to work on this. Now that branch is merged so basically PR is against master (old source branch doesn't exists anymore). I'll try to pull from master maybe probably the diff will become more effective, or I can create a new PR cherry picking only relevant commits from this one.
> 
> Just a merge and the common commits should disappear from this PR

Yep it's just 6 files now

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/trackedentity/AbstractTrackedEntityInstanceService.java (hunk 4 above)
Comment (gnespolino) [reply dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/trackedentity/AbstractTrackedEntityInstanceService.java]: I guess so, I will add it

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/DefaultProgramInstanceService.java [hunk 5]
```diff
@@ -261,7 +263,7 @@ public ProgramInstanceQueryParams getFromUrl( Set<String> ou, OrganisationUnitSe
         params.setProgramStartDate( programStartDate );
         params.setProgramEndDate( programEndDate );
         params.setTrackedEntityType( te );
-        params.setTrackedEntityInstanceUid( tei.getUid() );
+        params.setTrackedEntityInstanceUid(Optional.ofNullable(tei).map(BaseIdentifiableObject::getUid).orElse(null));
```
[the remark this answers is not in the document]
Comment (gnespolino) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/DefaultProgramInstanceService.java]: fixed

File: dhis-2/dhis-services/dhis-service-dxf2/src/test/java/org/hisp/dhis/dxf2/events/aggregates/TrackedEntityInstanceAggregateTest.java [hunk 6]
```diff
@@ -97,17 +97,17 @@ public void testFetchTrackedEntityInstances()
...
-            .getTrackedEntityInstances2( queryParams, params, false );
+            .getTrackedEntityInstances( queryParams, params, false );
 
         assertThat( trackedEntityInstances, hasSize( 4 ) );
         assertThat( trackedEntityInstances.get( 0 ).getEnrollments(), hasSize( 0 ) );
 
         //Check further for explicit uid in param
         queryParams.getTrackedEntityInstanceUids()
-            .addAll( trackedEntityInstances.stream().limit( 2 ).map( t -> t.getTrackedEntityInstance() ).collect( Collectors.toSet() ) );
+            .addAll( trackedEntityInstances.stream().limit( 2 ).map(TrackedEntityInstance::getTrackedEntityInstance).collect( Collectors.toSet() ) );
```
[the remark this answers is not in the document]
Comment (gnespolino) [reply dhis-2/dhis-services/dhis-service-dxf2/src/test/java/org/hisp/dhis/dxf2/events/aggregates/TrackedEntityInstanceAggregateTest.java]: fixed

File: dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/service/ContextService.java (hunk 2 above)
Comment (gnespolino) [reply dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/service/ContextService.java]: fixed

File: dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/service/TrackedEntityInstanceSupportService.java (hunk 3 above)
Comment (gnespolino) [reply dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/service/TrackedEntityInstanceSupportService.java]: > package declaration before copyright

fixed

**Review by enricocolasante — APPROVED** [review]

**Pull request merged**